python/ray/rllib/utils/sampler.py

```diff
@@ -0,0 +1,277 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import six.moves.queue as queue
+import threading
+from collections import namedtuple
+
+
+def lock_wrap(func, lock):
+    def wrapper(*args, **kwargs):
+        with lock:
+            return func(*args, **kwargs)
+    return wrapper
+
+
+class PartialRollout(object):
+    """A piece of a complete rollout.
+
+    We run our agent, and process its experience once it has processed enough
+    steps.
+    """
+
+    fields = ["state", "action", "reward", "terminal", "features"]
+
+    def __init__(self, extra_fields=None):
+        """Initializers internals. Maintains a `last_r` field
+        in support of partial rollouts, used in bootstrapping advantage
+        estimation.
+
+        Args:
+            extra_fields: Optional field for object to keep track.
+        """
+        if extra_fields:
+            self.fields.extend(extra_fields)
+        self.data = {k: [] for k in self.fields}
+        self.last_r = 0.0
+
+    def add(self, **kwargs):
+        for k, v in kwargs.items():
+            self.data[k] += [v]
+
+    def extend(self, other_rollout):
+        """Extends internal data structure. Assumes other_rollout contains
+        data that occured afterwards."""
+
+        assert not self.is_terminal()
+        assert all(k in other_rollout.fields for k in self.fields)
+        for k, v in other_rollout.data.items():
+            self.data[k].extend(v)
+        self.last_r = other_rollout.last_r
+
+    def is_terminal(self):
+        """Check if terminal.
+
+        Returns:
+            terminal (bool): if rollout has terminated."""
+        return self.data["terminal"][-1]
+
+
+CompletedRollout = namedtuple(
+    "CompletedRollout", ["episode_length", "episode_reward"])
+
+
+class SyncSampler(object):
```

Comment:
Does this and AsyncSampler implement any common interface?

Reply:
Yeah, should we make a class for that? I feel like there won't be any other Sampler classes though.